lib/spack/spack/config.py

```diff
@@ -999,6 +1028,40 @@ def default_list_scope():
     return None
 
 
+def _update_in_memory(data, section):
+    """Update the format of the configuration data in memory.
+
+    This function assumes the section is valid (i.e. validation
+    is responsibility of the caller)
+
+    Args:
+        data (dict): configuration data
+        section (str): section of the configuration to update
+
+    Returns:
+        True if the data was changed, False otherwise
+    """
+    update_fn = ensure_latest_format_fn(section)
+    changed = update_fn(data[section])
+    return changed
+
+
+def ensure_latest_format_fn(section):
+    """Return a function that takes as input a dictionary read from
+    a configuration file and update it to the latest format.
+
+    The function returns True if there was any update, False otherwise.
+
+    Args:
+        section (str): section of the configuration e.g. "packages",
+            "config", etc.
+    """
+    section_module_name = 'spack.schema.' + section
```

Comment:
https://github.com/spack/spack/pull/16526#discussion_r451757104 still applies here, although the code has since moved. Since the dictionary references the modules you can simplify like

```
section module = section_schemas[section]
update_fn = ...
```

Reply:
I don't think I can do that. The dictionary has schemas not module objects. I can though simplify the code by assuming all the modules I need have already been imported at the top level. This will result in:
```python
section_module = getattr(spack.schema, section)
```

Reply:
Done in 54f6702